extern/boostd-data/couchbase/db.go

```diff
@@ -665,3 +666,155 @@ func serviceName(svc gocb.ServiceType) string {
 	}
 	return "unknown"
 }
+
+// RemoveMetadata
+func (db *DB) RemoveMetadata(ctx context.Context, pieceCid cid.Cid) error {
+	ctx, span := tracing.Tracer.Start(ctx, "db.remove_metadata")
+	defer span.End()
+
+	return db.withCasRetry("remove-metadata-for-piece", func() error {
+		var getResult *gocb.GetResult
+		k := toCouchKey(pieceCid.String())
+		getResult, err := db.pcidToMeta.Get(k, &gocb.GetOptions{Context: ctx})
+		if err != nil {
+			return fmt.Errorf("getting piece cid to metadata for piece %s: %w", pieceCid, err)
+		}
+
+		var metadata CouchbaseMetadata
+		err = getResult.Content(&metadata)
+		if err != nil {
+			return fmt.Errorf("getting piece cid to metadata content for piece %s: %w", pieceCid, err)
+		}
+
+		// Remove all multihashes first, as without Metadata, they cannot be removed.
+		// This order is important as metadata.BlockCount is required in case RemoveIndexes fails
+		// and needs to be run manually
+		if err = db.RemoveIndexes(ctx, pieceCid, metadata.BlockCount); err != nil {
+			return fmt.Errorf("failed removing index for piece %s: %w", pieceCid, err)
+		}
+
+		_, err = db.pcidToMeta.Remove(k, &gocb.RemoveOptions{
+			Context: ctx,
+			Cas:     getResult.Cas(),
+		})
+		if err != nil {
+			return fmt.Errorf("removing piece %s metadata: %w", pieceCid, err)
+		}
+		return nil
+	})
+}
+
+// RemoveIndexes
+func (db *DB) RemoveIndexes(ctx context.Context, pieceCid cid.Cid, recordCount int) error {
+	ctx, span := tracing.Tracer.Start(ctx, "db.remove_indexes")
+	defer span.End()
+
+	// Get the number of shards
+	_, totalShards := getShardPrefixBitCount(recordCount)
+	for i := 0; i < totalShards; i++ {
+		// Get the map of multihash -> offset/size for the shard
+		shardPrefix, err := getShardPrefix(i)
+		if err != nil {
+			return err
+		}
+		cbKey := toCouchKey(pieceCid.String() + shardPrefix)
+		cbMap := db.pieceOffsets.Map(cbKey)
+		recMap, err := cbMap.Iterator()
+		if err != nil {
+			if isNotFoundErr(err) {
+				// If there are no records in a particular shard just skip the shard
+				continue
+			}
+			return fmt.Errorf("getting all records for piece %s: %w", pieceCid, err)
+		}
+
+		// Get each value in the map
+		for mhStr := range recMap {
+			cbKey := toCouchKey(mhStr)
+			err = db.withCasRetry("remove-piece", func() error {
+				getResult, err := db.mhToPieces.Get(cbKey, &gocb.GetOptions{Context: ctx})
+				if err != nil {
+					return nil
+				}
```

Comment:
If the key is not found, let's assume another process removed it, and continue (skip this key)

Reply:
This is wrapped in a cas retry func() so, putting a continue statement won't work. I can return the original error here and check the error outside the func() and then continue without calling the remove. WDYT?

Reply:
I think it makes more sense to check if the error is a not found error, and in that case return nil
Then withCasRetry will also return nil.
We should also check the error returned from db.pieceOffsets.Remove and if it's a not found error, ignore it.